### PR TITLE
CDVD: Added cdvdTrack, cdvdTrackIndex, assigned subQ from hardware reads

### PR DESCRIPTION
**Pull request by weirdbeardgame**
### Description of Changes

Added cdvdTrack, cdvdTrackIndex, changed sub Q mode field to ADR

### Rationale behind Changes

- Prep work for track-based formats like CHD, BIN / CUE. Using cdvdTrack as a common type for all track reading formats including CDVD Disc reader. 

- [Other than that, moar accuracy](https://psx-spx.consoledev.net/cdromdrive/#subchannel-q) 

### Suggested Testing Steps

Already tested, PR makes no significant changes to CD-based games outside of more correct SubQ reads and fixes Track layout in array. This fixes Dance Factory by pushing tracks into, 0 indexed array instead of 1 indexed

### CODE REVIEW

File: pcsx2/CDVD/Windows/IOCtlSrc.cpp [hunk 1]
```diff
@@ -303,6 +303,36 @@ bool IOCtlSrc::ReadCDInfo()
...
+	format.Format = IOCTL_CDROM_CURRENT_POSITION;
+
+	if (!DeviceIoControl(m_device, IOCTL_CDROM_READ_Q_CHANNEL, &format, sizeof(format), &osSubQ, sizeof(osSubQ), &unused, nullptr))
+	{
+		Console.Error("Subchannel Q reads failed");
+		return false;
+	}
+	else
+	{
+        subQ->adr = osSubQ.CurrentPosition.ADR;
```
Comment (F0bes) [inline pcsx2/CDVD/Windows/IOCtlSrc.cpp:321]: You're using spaces here

File: pcsx2/CDVD/Windows/IOCtlSrc.cpp [hunk 2]
```diff
@@ -303,6 +303,36 @@ bool IOCtlSrc::ReadCDInfo()
...
+		subQ->ctrl = osSubQ.CurrentPosition.Control;
+		subQ->trackNum = osSubQ.CurrentPosition.TrackNumber;
+		subQ->trackIndex = osSubQ.CurrentPosition.IndexNumber;
+		
+        subQ->trackM = osSubQ.CurrentPosition.TrackRelativeAddress[0];
```
Comment (F0bes) [inline pcsx2/CDVD/Windows/IOCtlSrc.cpp:326]: You're using spaces here

File: pcsx2/CDVD/Linux/IOCtlSrc.cpp [hunk 3]
```diff
@@ -194,6 +195,35 @@ bool IOCtlSrc::ReadCDInfo()
...
+
+	osSubQ.cdsc_format = CDROM_MSF;
+
+	if (ioctl(m_device, CDROMSUBCHNL, &osSubQ) == -1)
+	{
+		Console.Error("SUB CHANNEL READ ERROR: %s\n", strerror(errno));
+		return false;
+	}
+
+    subQ->adr = osSubQ.cdsc_adr;
```
Comment (F0bes) [inline pcsx2/CDVD/Linux/IOCtlSrc.cpp:210]: You're using spaces here

File: pcsx2/CDVD/Windows/IOCtlSrc.cpp [hunk 4]
```diff
@@ -303,6 +303,36 @@ bool IOCtlSrc::ReadCDInfo()
...
+		subQ->trackS = osSubQ.CurrentPosition.TrackRelativeAddress[1];
+		subQ->trackF = osSubQ.CurrentPosition.TrackRelativeAddress[2];
+
+		subQ->discM = osSubQ.CurrentPosition.AbsoluteAddress[0];
+		subQ->discS = osSubQ.CurrentPosition.AbsoluteAddress[1];
+		subQ->discF = osSubQ.CurrentPosition.AbsoluteAddress[2];
+	}
```
Comment (F0bes) [inline pcsx2/CDVD/Windows/IOCtlSrc.cpp:333]: You have no return here, this is UB. Are you sure you tested on Windows?

File: pcsx2/CDVD/Darwin/IOCtlSrc.cpp [hunk 5]
```diff
@@ -241,6 +241,11 @@ bool IOCtlSrc::ReadCDInfo()
 #endif
 }
 
+bool IOCtlSrc::ReadTrackSubQ(cdvdSubQ* subQ) const
+{
+    return false;
```
Comment (F0bes) [inline pcsx2/CDVD/Darwin/IOCtlSrc.cpp:246]: You're using spaces here

File: pcsx2/CDVD/CDVDdiscReader.cpp [hunk 6]
```diff
@@ -4,6 +4,7 @@
 #include "CDVDdiscReader.h"
```
Comment (F0bes) [inline pcsx2/CDVD/CDVDdiscReader.cpp:4]: Please format this file.

File: pcsx2/CDVD/CDVDdiscReader.cpp [hunk 7]
```diff
@@ -270,20 +271,41 @@ static s32 DISCreadSubQ(u32 lsn, cdvdSubQ* subq)
 
 	memset(subq, 0, sizeof(cdvdSubQ));
 
-	lsn_to_msf(&subq->discM, &subq->discS, &subq->discF, lsn + 150);
+	if (src->ReadTrackSubQ(subq) != 0)
```
Comment (F0bes) [inline pcsx2/CDVD/CDVDdiscReader.cpp:274]: ReadTrackSubQ returns a boolean, please compare as so.

File: pcsx2/CDVD/CDVDdiscReader.cpp [hunk 8]
```diff
@@ -270,20 +271,41 @@ static s32 DISCreadSubQ(u32 lsn, cdvdSubQ* subq)
...
+	{
+		u8 i = strack;
+		while (i < etrack && lsn >= tracks[i + 1].start_lba)
+			++i;
+
+		lsn -= tracks[i].start_lba;
+
+		/*if (DATA(subq->ctrl))
```
Comment (F0bes) [inline pcsx2/CDVD/CDVDdiscReader.cpp:282]: Remove code if it is not used. If you feel like it is important to a developer to know that DATA(subq->ctrl) is set for data, present it as a comment on that field.

**Review by F0bes — CHANGES_REQUESTED** [review]

File: pcsx2/CDVD/CDVDdiscReader.cpp [hunk 9]
```diff
@@ -270,20 +271,32 @@ static s32 DISCreadSubQ(u32 lsn, cdvdSubQ* subq)
 
 	memset(subq, 0, sizeof(cdvdSubQ));
 
-	lsn_to_msf(&subq->discM, &subq->discS, &subq->discF, lsn + 150);
+	if (src->ReadTrackSubQ(subq))
```
[the remark this answers is not in the document]
Comment (weirdbeardgame) [reply pcsx2/CDVD/CDVDdiscReader.cpp]: I don't know if we need to or not. I do know, that outside of this, the "subQ" is cast to `SCMDResultBuff[1]` in [CDVD.cpp:2362](https://github.com/PCSX2/pcsx2/blob/5c83edb7b2ab84afa4e423e889c278f6df363230/pcsx2/CDVD/CDVD.cpp#L2362) So perhaps that's what does it.

On the other end of that spectrum, the bios CD player is currently in master broke in that it'll read a single track frame then just ... Stop. I don't know if that has anything to do with it.

**Comment by weirdbeardgame** [comment]
Btw, if anyone wants to know more about the Subchannel IOCtrl structs

## Linux:
https://docs.kernel.org/userspace-api/ioctl/cdrom.html
https://github.com/torvalds/linux/blob/77f587896757708780a7e8792efe62939f25a5ab/include/uapi/linux/cdrom.h#L208

## Windows:
https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntddcdrm/ns-ntddcdrm-_sub_q_channel_data
https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntddcdrm/ns-ntddcdrm-_sub_q_current_position
https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntddcdrm/ns-ntddcdrm-_cdrom_sub_q_data_format

**Comment by weirdbeardgame** [comment]
Just pushed the commit that fixed Dance Factory.

File: pcsx2/CDVD/CDVDcommon.h [hunk 10]
```diff
@@ -65,6 +94,12 @@ typedef struct _cdvdTN
 #define CDVD_TYPE_DETCT 0x01 // Detecting
 #define CDVD_TYPE_NODISC 0x00 // No Disc
 
+// SUBQ CONTROL:
+#define AUDIO_PREEMPHASIS(control) ((control & (4 << 1)))
```
Comment (F0bes) [inline pcsx2/CDVD/CDVDcommon.h:98]: If these are intended to be used as true/false helpers. I would rather them be named stuff like CDVD_IS_DIGITAL_COPY, CDVD_IS_DATA, CDVD_HAS_QUADRAPHONIC_AUDIO, or whatever makes sense but with that intent.

It makes the usage of those a little more clear and it also doesn't populate any files that include this file (this is a common file) with a bunch of macros without a namespace.

**Review by F0bes — CHANGES_REQUESTED** [review]

**Comment by F0bes** [comment]
I'm awaiting the arrival of my DVD drive and I'll be hardware testing this.

**Comment by TheLastRar** [comment]
The second commit needs to be prefixed (presumably with CDVD:)

**Comment by weirdbeardgame** [comment]
done

**Comment by F0bes** [comment]
Certainly an improvement to master.
There seems to be some other info missing from bytes 4 5 and 6 for each track (diff the results). I don't know if that could be in the scope of this PR?

<details>
<summary>PS2 Results</summary>

```
Number of tracks: 14
Track 1: 150 (type:1)
        00:02:00
        (01 00 01 02 30 36 00 00 02 00)
Track 2: 20442 (type:1)
        04:32:42
        (01 00 02 02 30 39 00 04 32 42)
Track 3: 40275 (type:1)
        08:57:00
        (01 00 03 02 30 42 00 08 57 00)
Track 4: 57852 (type:1)
        12:51:27
        (01 00 04 02 30 45 00 12 51 27)
Track 5: 71240 (type:1)
        15:49:65
        (01 00 05 02 30 48 00 15 49 65)
Track 6: 99225 (type:1)
        22:03:00
        (01 00 06 02 30 51 00 22 03 00)
Track 7: 121637 (type:1)
        27:01:62
        (01 00 07 02 30 54 00 27 01 62)
Track 8: 146492 (type:1)
        32:33:17
        (01 00 08 02 30 57 00 32 33 17)
Track 9: 172095 (type:1)
        38:14:45
        (01 00 09 02 30 60 00 38 14 45)
Track 10: 179315 (type:1)
        39:50:65
        (01 00 10 02 30 63 00 39 50 65)
Track 11: 194510 (type:1)
        43:13:35
        (01 00 11 02 30 66 00 43 13 35)
Track 12: 216522 (type:1)
        48:06:72
        (01 00 12 02 30 67 00 48 06 72)
Track 13: 247447 (type:1)
        54:59:22
        (01 00 13 02 30 21 00 54 59 22)
Track 14: 265215 (type:1)
        58:56:15
        (01 00 14 02 30 24 00 58 56 15)
```
</details>

<details>
<summary>PS2 (PR) Results</summary>

```
Number of tracks: 14
Track 1: 0 (type:0)
        00:00:00
        (00 00 00 00 00 00 00 00 00 00)
Track 2: 20442 (type:1)
        04:32:42
        (01 00 01 00 00 00 00 04 32 42)
Track 3: 40275 (type:1)
        08:57:00
        (01 00 02 00 00 00 00 08 57 00)
Track 4: 57852 (type:1)
        12:51:27
        (01 00 03 00 00 00 00 12 51 27)
Track 5: 71240 (type:1)
        15:49:65
        (01 00 04 00 00 00 00 15 49 65)
Track 6: 99225 (type:1)
        22:03:00
        (01 00 05 00 00 00 00 22 03 00)
Track 7: 121637 (type:1)
        27:01:62
        (01 00 06 00 00 00 00 27 01 62)
Track 8: 146492 (type:1)
        32:33:17
        (01 00 07 00 00 00 00 32 33 17)
Track 9: 172095 (type:1)
        38:14:45
        (01 00 08 00 00 00 00 38 14 45)
Track 10: 179315 (type:1)
        39:50:65
        (01 00 09 00 00 00 00 39 50 65)
Track 11: 194510 (type:1)
        43:13:35
        (01 00 10 00 00 00 00 43 13 35)
Track 12: 216522 (type:1)
        48:06:72
        (01 00 11 00 00 00 00 48 06 72)
Track 13: 247447 (type:1)
        54:59:22
        (01 00 12 00 00 00 00 54 59 22)
Track 14: 265215 (type:1)
        58:56:15
        (01 00 13 00 00 00 00 58 56 15)
```
</details>

<details>
<summary>PCSX2 (MASTER) Results</summary>

```

Number of tracks: 14
Track 1: 0 (type:0)
        00:00:00
        (00 00 00 00 00 00 00 00 00 00)
Track 2: 150 (type:1)
        00:02:00
        (01 00 01 00 00 00 00 00 02 00)
Track 3: 20442 (type:1)
        04:32:42
        (01 00 02 00 00 00 00 04 32 42)
Track 4: 40275 (type:1)
        08:57:00
        (01 00 03 00 00 00 00 08 57 00)
Track 5: 57852 (type:1)
        12:51:27
        (01 00 04 00 00 00 00 12 51 27)
Track 6: 71240 (type:1)
        15:49:65
        (01 00 05 00 00 00 00 15 49 65)
Track 7: 99225 (type:1)
        22:03:00
        (01 00 06 00 00 00 00 22 03 00)
Track 8: 121637 (type:1)
        27:01:62
        (01 00 07 00 00 00 00 27 01 62)
Track 9: 146492 (type:1)
        32:33:17
        (01 00 08 00 00 00 00 32 33 17)
Track 10: 172095 (type:1)
        38:14:45
        (01 00 09 00 00 00 00 38 14 45)
Track 11: 179315 (type:1)
        39:50:65
        (01 00 10 00 00 00 00 39 50 65)
Track 12: 194510 (type:1)
        43:13:35
        (01 00 11 00 00 00 00 43 13 35)
Track 13: 216522 (type:1)
        48:06:72
        (01 00 12 00 00 00 00 48 06 72)
Track 14: 247447 (type:1)
        54:59:22
        (01 00 13 00 00 00 00 54 59 22)
```

</details>

<details>
<summary>cdparanoia</summary>

```
Table of contents (audio tracks only):
track        length               begin        copy pre ch
===========================================================
  1.    20292 [04:30.42]        0 [00:00.00]    no   no  2
  2.    19833 [04:24.33]    20292 [04:30.42]    no   no  2
  3.    17577 [03:54.27]    40125 [08:55.00]    no   no  2
  4.    13388 [02:58.38]    57702 [12:49.27]    no   no  2
  5.    27985 [06:13.10]    71090 [15:47.65]    no   no  2
  6.    22412 [04:58.62]    99075 [22:01.00]    no   no  2
  7.    24855 [05:31.30]   121487 [26:59.62]    no   no  2
  8.    25603 [05:41.28]   146342 [32:31.17]    no   no  2
  9.     7220 [01:36.20]   171945 [38:12.45]    no   no  2
 10.    15195 [03:22.45]   179165 [39:48.65]    no   no  2
 11.    22012 [04:53.37]   194360 [43:11.35]    no   no  2
 12.    30925 [06:52.25]   216372 [48:04.72]    no   no  2
 13.    17768 [03:56.68]   247297 [54:57.22]    no   no  2
 14.    28055 [06:14.05]   265065 [58:54.15]    no   no  2
TOTAL  293120 [65:08.20]    (audio only)
```

</details>

**Review by F0bes — APPROVED** [review]
As I said in my comment this brings master closer to a real console.
I tested both Linux and Windows and they produce an identical result.

**Pull request merged**